### PR TITLE
Reduce memory allocations in C++ to PyList Conversions

### PR DESCRIPTION
**Pull request by maxbachmann**

### CODE REVIEW

File: Cython/Utility/CppConvert.pyx [hunk 1]
```diff
@@ -52,14 +52,24 @@ cdef vector[X] {{cname}}(object o) except *:
 #################### vector.to_py ####################
 
 cdef extern from *:
-    cdef cppclass vector "const std::vector" [T]:
+    cdef cppclass vector "std::vector" [T]:
```
Comment (maxbachmann) [inline Cython/Utility/CppConvert.pyx:55]: In all other cases const is added in the function signature instead. Is there a reason, why it has to be in here for std::vector?

File: Cython/Utility/CppConvert.pyx [hunk 2]
```diff
@@ -52,14 +52,24 @@ cdef vector[X] {{cname}}(object o) except *:
...
 
+@cname("{{cname}}")
+cdef object {{cname}}(const vector[X]& v):
+    o = PyList_New(v.size())
+    cdef size_t i = 0
+    for i in range(v.size()):
+        item = v[i]
+        Py_INCREF(item)
+        PyList_SET_ITEM(o, i, item)
+    return o
```
Comment (maxbachmann) [inline Cython/Utility/CppConvert.pyx:72]: This applies changes similar to https://github.com/cython/cython/issues/2844, since it is not really clear when/if there will be such a optimization for list comprehensions. In case it exists in the future this could directly be a list comprehension as well

File: Cython/Utility/CppConvert.pyx (hunk 1 above)
Comment (scoder) [reply Cython/Utility/CppConvert.pyx:55]: Probably just predates proper `const` support.

File: Cython/Utility/CppConvert.pyx [hunk 3]
```diff
@@ -52,14 +52,24 @@ cdef vector[X] {{cname}}(object o) except *:
...
-    return [v[i] for i in range(v.size())]
+cdef extern from "Python.h":
+    void Py_INCREF(object)
+    list PyList_New(Py_ssize_t size)
+    void PyList_SET_ITEM(object list, Py_ssize_t i, object o)
 
+@cname("{{cname}}")
+cdef object {{cname}}(const vector[X]& v):
+    o = PyList_New(v.size())
+    cdef size_t i = 0
```
Comment (scoder) [inline Cython/Utility/CppConvert.pyx:67]: Does this need to be `size_t` instead of the usual `Py_ssize_t` for indexing?
```suggestion
    cdef Py_ssize_t i
```

File: Cython/Utility/CppConvert.pyx [hunk 4]
```diff
@@ -87,14 +97,24 @@ cdef extern from *:
...
+    list PyList_New(Py_ssize_t size)
+    void PyList_SET_ITEM(object list, Py_ssize_t i, object o)
 
 @cname("{{cname}}")
 cdef object {{cname}}(const cpp_list[X]& v):
-    o = []
-    cdef cpp_list[X].const_iterator iter = v.begin()
+    o = PyList_New(v.size())
+    cdef size_t i = 0
+    cdef cpp_list[int].const_iterator iter = v.begin()
```
Comment (scoder) [inline Cython/Utility/CppConvert.pyx:111]: `int`?
```suggestion
    cdef cpp_list[X].const_iterator iter = v.begin()
```

File: Cython/Utility/CppConvert.pyx [hunk 5]
```diff
@@ -87,14 +97,24 @@ cdef extern from *:
...
     while iter != v.end():
-        o.append(cython.operator.dereference(iter))
+        item = cython.operator.dereference(iter)
+        Py_INCREF(item)
+        PyList_SET_ITEM(o, i, item)
         cython.operator.preincrement(iter)
+        i = i + 1
```
Comment (scoder) [inline Cython/Utility/CppConvert.pyx:117]:
```suggestion
        i += 1
```

File: Cython/Utility/CppConvert.pyx (hunk 3 above)
Comment (da-woods) [reply Cython/Utility/CppConvert.pyx:67]: This might just be a mismatch where no option can ever be "right" - `size_t` matches the type of of `v.size()`, but not `PyList_SET_ITEM`.

Possibly the right thing to do is to raise an error if `v.size() > PY_SSIZE_T_MAX` and then use `Py_ssize_t` after that check? But I doubt if many people have the memory to run that test-case.

File: Cython/Utility/CppConvert.pyx [hunk 6]
```diff
@@ -52,14 +52,24 @@ cdef vector[X] {{cname}}(object o) except *:
...
+cdef extern from "Python.h":
+    void Py_INCREF(object)
+    list PyList_New(Py_ssize_t size)
+    void PyList_SET_ITEM(object list, Py_ssize_t i, object o)
 
+@cname("{{cname}}")
+cdef object {{cname}}(const vector[X]& v):
+    o = PyList_New(<Py_ssize_t>v.size())
+    cdef Py_ssize_t i = 0
+    for i in range(v.size()):
```
Comment (scoder) [inline Cython/Utility/CppConvert.pyx:68]: Useless assignment.
```suggestion
    cdef Py_ssize_t i
    for i in range(v.size()):
```

File: Cython/Utility/CppConvert.pyx (hunk 3 above)
Comment (maxbachmann) [reply Cython/Utility/CppConvert.pyx]: I changed it to Py_ssize_t. Even for a bit vector (vector<bool>) this would require 1.07E+09 Gb of memory, so I do not think, that this is a concern (+ the memory allocation for the PyList). In Cpython this would already causes a exception because a negative size is specified, while in PyPy this would allocate a empty list.
https://github.com/mozillazg/pypy/blob/11af013b3723c15c9ce51f68fd5c2db51f74ba3d/pypy/module/cpyext/listobject.py#L23
I personally do not see anyone running into this issue, given the big memory requirement

File: Cython/Utility/CppConvert.pyx [hunk 7]
```diff
@@ -52,14 +52,24 @@ cdef vector[X] {{cname}}(object o) except *:
...
-cdef object {{cname}}(vector[X]& v):
-    return [v[i] for i in range(v.size())]
+cdef extern from "Python.h":
+    void Py_INCREF(object)
+    list PyList_New(Py_ssize_t size)
+    void PyList_SET_ITEM(object list, Py_ssize_t i, object o)
 
+@cname("{{cname}}")
+cdef object {{cname}}(const vector[X]& v):
+    o = PyList_New(<Py_ssize_t>v.size())
```
Comment (scoder) [inline Cython/Utility/CppConvert.pyx:66]: It still feels like we should add a max size check here (and also for `cpp_list` further down), because, why not?

File: Cython/Utility/CppConvert.pyx (hunk 6 above)
Comment (maxbachmann) [reply Cython/Utility/CppConvert.pyx]: Is it preferred to use
```
    cdef Py_ssize_t i
    for i in range(v.size()):
```
or
```
    for i in range(<Py_ssize_t>v.size()):
```
and let cython deduce the type of i?

File: Cython/Utility/CppConvert.pyx [hunk 8]
```diff
@@ -52,14 +52,24 @@ cdef vector[X] {{cname}}(object o) except *:
...
+    cdef Py_ssize_t i
+    for i in range(v.size()):
+        item = v[i]
+        Py_INCREF(item)
+        PyList_SET_ITEM(o, i, item)
```
Comment (scoder) [inline Cython/Utility/CppConvert.pyx:71]: I think we should add a size check here to make sure that the vector did not grow during iteration.

File: Cython/Utility/CppConvert.pyx [hunk 9]
```diff
@@ -87,14 +97,24 @@ cdef extern from *:
...
 cdef object {{cname}}(const cpp_list[X]& v):
-    o = []
+    o = PyList_New(<Py_ssize_t>v.size())
+    cdef Py_ssize_t i = 0
     cdef cpp_list[X].const_iterator iter = v.begin()
     while iter != v.end():
-        o.append(cython.operator.dereference(iter))
+        item = cython.operator.dereference(iter)
+        Py_INCREF(item)
+        PyList_SET_ITEM(o, i, item)
```
Comment (scoder) [inline Cython/Utility/CppConvert.pyx:115]: I think we should add a size check here to make sure that the vector did not grow during iteration.

File: Cython/Utility/CppConvert.pyx (hunk 6 above)
Comment (scoder) [reply Cython/Utility/CppConvert.pyx]: The first. It's usually better if utility code doesn't depend too much on features that users can influence. Keeps it less brittle.

File: Cython/Utility/CppConvert.pyx (hunk 8 above)
Comment (maxbachmann) [reply Cython/Utility/CppConvert.pyx:71]: Can this happen? Since this requires the GIL I do not see how a different thread would change the size + generally C++ standard containers are not thread safe.

File: Cython/Utility/CppConvert.pyx (hunk 7 above)
Comment (maxbachmann) [reply Cython/Utility/CppConvert.pyx:66]: what is the appropriate exception type?

File: Cython/Utility/CppConvert.pyx (hunk 8 above)
Comment (da-woods) [reply Cython/Utility/CppConvert.pyx:71]: In principle the constructor/destructor of any of the C++ objects could resize the vector, or the the constructor of the `item` Python object could release the GIL (most likely by invoking the GC, and causing arbitrary Python code to get executed in a destructor somewhere).

Personally I'd take the view that pretty much all of C++ iteration relies on containers not managing to resize themselves during it, and that it isn't necessary to handle this corner case. But I'm pretty sure I could force it to happen if I tried.

Comment (da-woods) [reply Cython/Utility/CppConvert.pyx:71]: With that said, if we were going to test it I think the place to do it would be before indexing the vector - I don't think that the vector growing is a problem for the current code because the stop condition for the range is only evaluated once

Comment (scoder) [reply Cython/Utility/CppConvert.pyx:71]: Right. This case is probably ok. If the vector grows, it will stop short, which is fine since we don't guarantee anything in that case except that it doesn't crash. If the vector shrinks, it would raise a C++ exception on indexing, which should then get turned into an `IndexError`. So that's fine, too.

File: Cython/Utility/CppConvert.pyx (hunk 7 above)
Comment (scoder) [reply Cython/Utility/CppConvert.pyx:66]: `OverflowError`?

File: Cython/Utility/CppConvert.pyx (hunk 8 above)
Comment (maxbachmann) [reply Cython/Utility/CppConvert.pyx:71]: Yes probably you could force it. C++ containers are not thread safe, so users should manage this themselves. I do not see any way for us to make sure they are in this use case. Checking the size in front of each access would just make it harder to break it, but if you try hard you can still break it.

Comment (maxbachmann) [reply Cython/Utility/CppConvert.pyx:71]: > If the vector shrinks, it would raise a C++ exception on indexing, which should then get turned into an IndexError. So that's fine, too.
No it does not
v[i] just accesses out of bouns memory which is undefined behavior

v.at(i) raises an exception, but since vector is not thread safe it might resize in between the size check and the access in the at function

Also many of our iterations are based on iterators, which might get invalidated -> undefined behavior

Comment (scoder) [reply Cython/Utility/CppConvert.pyx:71]: I think threading is not something we should care about. But modification during iteration is. As @da-woods pointed out, the vector might change while instantiating the Python item, which can execute arbitrary Python code. We should guard against that.

Comment (scoder) [reply Cython/Utility/CppConvert.pyx:71]: We could use a `while` loop and add a second loop after it that uses `append()` if there is really anything left to append.

Comment (da-woods) [reply Cython/Utility/CppConvert.pyx:71]: I'm not sure a second loop is worthwhile. I'd argue for either:
* Use the code as-is (on the basis this anything that changes the vector during iteration is very unusual, and people who write such code should be ultra-careful about ownership and object lifetime)
* Swap `[i]` for `.at(I)`, on the basis that it should keep things fairly safe (i.e. raise an Python exception)

Personally I don't think it's sensible to try to handle too much beyond that. Any code that could change the size of vector could also go back and modify earlier elements, so we can never do more than just a "best-effort snapshot"

Comment (scoder) [reply Cython/Utility/CppConvert.pyx:71]: I would argue that we should mimic the semantics of Python list iteration here. If you change a Python list while iterating over it, it will simply stop as soon as it has returned as many (or more) items as the current list length at this time. It may skip over items when you add or remove some, but it won't crash, and it will look at the list length on every iteration.

Note that this is not special-purpose code. It's generic list copying code that users won't even notice they are using most of the time. It should try to be as widely applicable as possible (but not a tad slower than strictly necessary). If we introduce potential bugs or non-obvious constraints here, then there is a high chance that it won't be noticed in user code until much, much later, when the right kind of data passes through it during the right phase of the moon.

Comment (da-woods) [reply Cython/Utility/CppConvert.pyx:71]: > Also many of our iterations are based on iterators, which might get invalidated -> undefined behavior

One final comment on this from me I think: for C++ linked lists, sets and maps, using iterators is the only way to traverse then and C++ iterators have a very strong requirement that the container won't be modified as you go. So for vector/string/deque we could definitely make it a bit more robust to changes, but for most other  types I don't think it's possible.

Comment (scoder) [reply Cython/Utility/CppConvert.pyx:71]: @da-woods That's a convincing reason for those containers then. :)

Just keep in mind that C++ containers can easily be used by Cython developers as low-level data types even if their C++ knowledge is limited. If we can make it work for them by providing sensible semantics, we should.

Comment (scoder) [reply Cython/Utility/CppConvert.pyx:71]: @da-woods What you wrote doesn't seem to apply to linked lists, though, e.g. https://www.cplusplus.com/reference/list/list/push_front/ – deques yes, maps and sets partly. Clearing always invalidates, for example, but inserting doesn't necessarily. So there are several modifications that users can apply to C++ containers without rendering their iterators invalid or useless.

Comment (maxbachmann) [reply Cython/Utility/CppConvert.pyx:71]: Yes linked lists are not that problematic. The iterator can simply check the next pointer each time. It is only problematic when the element the iterator currently points to is invalidated.

File: Cython/Utility/CppConvert.pyx (hunk 7 above)
Comment (scoder) [reply Cython/Utility/CppConvert.pyx:66]: Actually, raising a `MemoryError` might be better. After all, we'd be trying to allocate a huge list of Python objects (4-8 byte pointers plus objects), which would obviously fail if already its size exceeds the maximum allowed length. The size check would just prevent the size overflow that might lead to silent data loss instead of an exception.

Basically, the only realistic case where this may occur is a bit vector. Or maybe a `char` vector. Or something like that. Maybe something forged. But we should handle (and reject) such extreme cases correctly. Whenever we cast an unsigned to a signed size, it's worth a maximum size check.

**Review by scoder — COMMENTED** [review]
Would be nice to get the safety checks in so that this can be merged.

**Comment by maxbachmann** [comment]
@scoder I added the overflow check for the Py_ssize_t conversions

**Comment by maxbachmann** [comment]
Appears that I killed something. I will try to fix this later today.

**Comment by maxbachmann** [comment]
@scoder I just checked the generated functions and at least `__pyx_convert_vector_to_py_double` appears to pretty broken. See this full code:

```cpp
static PyObject *__pyx_convert_vector_to_py_double(std::vector<double>  const &__pyx_v_v) {
  PyObject *__pyx_v_o = NULL;
  Py_ssize_t __pyx_v_i;
  double __pyx_v_item;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  size_t __pyx_t_2;
  size_t __pyx_t_3;
  Py_ssize_t __pyx_t_4;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__pyx_convert_vector_to_py_double", 0);

  /* "vector.to_py":70
 * 
 * 
 *     o = PyList_New(<Py_ssize_t>v.size())             # <<<<<<<<<<<<<<
 *     cdef Py_ssize_t i
 *     for i in range(v.size()):
 */
  /* __pyx_t_1 allocated (Python object) */
  __pyx_t_1 = PyList_New(((Py_ssize_t)__pyx_v_v.size())); if (unlikely(!__pyx_t_1)) __PYX_ERR(1, 70, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_o = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;
  /* __pyx_t_1 released */

  /* "vector.to_py":72
 *     o = PyList_New(<Py_ssize_t>v.size())
 *     cdef Py_ssize_t i
 *     for i in range(v.size()):             # <<<<<<<<<<<<<<
 *         item = v[i]
 *         Py_INCREF(item)
 */
  /* __pyx_t_2 allocated (size_t) */
  __pyx_t_2 = __pyx_v_v.size();
  /* __pyx_t_3 allocated (size_t) */
  __pyx_t_3 = __pyx_t_2;
  /* __pyx_t_4 allocated (Py_ssize_t) */
  for (__pyx_t_4 = 0; __pyx_t_4 < __pyx_t_3; __pyx_t_4+=1) {
    __pyx_v_i = __pyx_t_4;

    /* "vector.to_py":73
 *     cdef Py_ssize_t i
 *     for i in range(v.size()):
 *         item = v[i]             # <<<<<<<<<<<<<<
 *         Py_INCREF(item)
 *         PyList_SET_ITEM(o, i, item)
 */
    __pyx_v_item = (__pyx_v_v[__pyx_v_i]);

    /* "vector.to_py":74
 *     for i in range(v.size()):
 *         item = v[i]
 *         Py_INCREF(item)             # <<<<<<<<<<<<<<
 *         PyList_SET_ITEM(o, i, item)
 *     return o
 */
    /* __pyx_t_1 allocated (Python object) */
    __pyx_t_1 = PyFloat_FromDouble(__pyx_v_item); if (unlikely(!__pyx_t_1)) __PYX_ERR(1, 74, __pyx_L1_error)
    __Pyx_GOTREF(__pyx_t_1);
    Py_INCREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    /* __pyx_t_1 released */

    /* "vector.to_py":75
 *         item = v[i]
 *         Py_INCREF(item)
 *         PyList_SET_ITEM(o, i, item)             # <<<<<<<<<<<<<<
 *     return o
 * 
 */
    /* __pyx_t_1 allocated (Python object) */
    __pyx_t_1 = PyFloat_FromDouble(__pyx_v_item); if (unlikely(!__pyx_t_1)) __PYX_ERR(1, 75, __pyx_L1_error)
    __Pyx_GOTREF(__pyx_t_1);
    PyList_SET_ITEM(__pyx_v_o, __pyx_v_i, __pyx_t_1);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    /* __pyx_t_1 released */
  }
  /* __pyx_t_4 released */
  /* __pyx_t_3 released */
  /* __pyx_t_2 released */

  /* "vector.to_py":76
 *         Py_INCREF(item)
 *         PyList_SET_ITEM(o, i, item)
 *     return o             # <<<<<<<<<<<<<<
 * 
 */
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_o);
  __pyx_r = __pyx_v_o;
  goto __pyx_L0;

  /* "vector.to_py":66
 * 
 * @cname("__pyx_convert_vector_to_py_double")
 * cdef object __pyx_convert_vector_to_py_double(const vector[X]& v):             # <<<<<<<<<<<<<<
 * 
 * 
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("vector.to_py.__pyx_convert_vector_to_py_double", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_o);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}
```
I am not sure why it:
1) creates the double objects
2) increfs it correctly
3) decrefs it and sets object to 0 again
4) creates the object again
5) sets the item in the list

I do not understand why steps 3 and 4 are generated.
Using
```
cdef extern from *:
    cdef cppclass vector "std::vector" [T]:
        size_t size()
        T& operator[](size_t)

cdef extern from "Python.h":
    void Py_INCREF(object)
    list PyList_New(Py_ssize_t size)
    void PyList_SET_ITEM(object list, Py_ssize_t i, object o)
    cdef Py_ssize_t PY_SSIZE_T_MAX

cdef object test_func(const vector[float]& v):
    if v.size() > PY_SSIZE_T_MAX:
        raise MemoryError()

    o = PyList_New(<Py_ssize_t>v.size())
    cdef Py_ssize_t i
    for i in range(v.size()):
        item = v[i]
        Py_INCREF(item)
        PyList_SET_ITEM(o, i, item)
    return o
```
in a standalone `.pyx` file generates the correct code. Is this a bug in cython, or is there something wrong with the implementation?

**Comment by da-woods** [comment]
In the line `item = v[i]` Cython is inferring the type of `item` as a double. Therefore `Py_INCREF(item)` translates as "convert `item` to a Python object, then incref that object". `PyList_SET_ITEM(o, i, item)` translates as "convert `item` to _another_ Python object and at that to the list.

I think `cdef object item` at the start of the function should fix it.

**Comment by maxbachmann** [comment]
Thanks @da-woods this fixed the problem :)

**Comment by scoder** [comment]
Thanks
